### PR TITLE
Handle SIGINT signals

### PR DESCRIPTION
**Pull request by Jackbaude**
Wait for a SIGINT signal from the process and clean up UDS socket before shutdown.

Signed-off-by: Jack <jackbaude@gmail.com>

### CODE REVIEW

**Comment by baude** [comment]
you need to run the validator

**Comment by Jackbaude** [comment]
@baude All checks passed. Please let me know thoughts on line 132 as mentioned above.

File: src/server.rs [hunk 1]
```diff
@@ -130,16 +145,17 @@ pub async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let uds_path = opts.uds.unwrap_or_else(|| DEFAULT_UDS_PATH.to_string());
     // timeout time if no leases are found
     let timeout = opts.timeout.unwrap_or(DEFAULT_TIMEOUT);
-
-    // Match because parent reruns an option
+    // Create a new uds socket path
     match Path::new(&uds_path).parent() {
         None => {
             log::error!("Could not find uds path");
```
Comment (baude) [inline src/server.rs:151]: error should have the fq path

File: src/server.rs [hunk 2]
```diff
@@ -118,6 +118,21 @@ struct Opts {
     timeout: Option<isize>,
 }
 
+/// Handle SIGINT signal.
+///
+/// Will wait until process receives a SIGINT/ ctrl+c signal and then clean up and shut down
+async fn handle_signal(uds_path: String) {
+    tokio::spawn(async move {
+        tokio::signal::ctrl_c()
```
[the remark this answers is not in the document]
Comment (Luap99) [reply src/server.rs:126]: you cannot handle sigkill, but we should handle sigterm as well

**Review by Luap99 — COMMENTED** [review]
Is the problem here to we cannot bind the socket when it already exists?

In any case we have to remove the socket path before calling bind since we cannot guarantee that the path was removed, i.e. SIGKILL or just a crash.

**Review by flouthoc — CHANGES_REQUESTED** [review]
You might wanna use https://docs.rs/tokio/latest/tokio/signal/unix/struct.SignalKind.html and directly process signal instead of top level implementation `ctrl_c` since there is no top level implementation for other signals.

Agree with @Luap99, this should handle `SIGTERM` as well

**Comment by flouthoc** [comment]
Maybe something on the lines of ( you can remove unwrap in actual code)
```rs
tokio::signal::unix::signal(SignalKind::terminate()).unwrap().recv().await
```

**Comment by baude** [comment]
please add a test

**Comment by Jackbaude** [comment]
@baude ptal

**Comment by Jackbaude** [comment]
@baude, @flouthoc, @Luap99. Is this PR ready to merge?

**Comment by baude** [comment]
did you resolve all the comments from folks @Jackbaude ?

File: src/server.rs [hunk 3]
```diff
@@ -130,6 +132,31 @@ struct Opts {
...
+///
+/// Will wait until process receives a SIGINT/ ctrl+c signal and then clean up and shut down
+async fn handle_signal(uds_path: String) {
+    tokio::spawn(async move {
+        // Handle signal hooks with expect, it is important these are setup so data is not corrupted
+        let mut sigterm = signal(SignalKind::terminate()).expect("Could not set up SIGTERM hook");
+        let mut sigint = signal(SignalKind::interrupt()).expect("Could not set up SIGINT hook");
+        // Wait for either a SIGINT or a SIGTERM to clean up
+        tokio::select! {
+            _term = sigterm.recv() => {
```
Comment (flouthoc) [inline src/server.rs:145]:
```suggestion
            _ = sigterm.recv() => {
```

File: src/server.rs [hunk 4]
```diff
@@ -130,6 +132,31 @@ struct Opts {
...
+                warn!("Received SIGTERM, cleaning up and exiting");
+            }
+            _int = sigint.recv() => {
```
Comment (flouthoc) [inline src/server.rs:148]:
```suggestion
            _ = sigint.recv() => {
```

File: src/server.rs [hunk 5]
```diff
@@ -130,6 +132,31 @@ struct Opts {
...
+                warn!("Received SIGINT, cleaning up and exiting");
+            }
+        }
+        if let Err(e) = fs::remove_file(uds_path) {
+            error!("Could not close uds socket: {}", e);
+        }
+
+        std::process::exit(0x0100);
```
Comment (flouthoc) [inline src/server.rs:156]: Is the reason for `0x0100`  to be `256` on windows or something else ?

**Comment by flouthoc** [comment]
Some non-blocking nits but overall LGTM

**Comment by Jackbaude** [comment]
> Some non-blocking nits but overall LGTM

@flouthoc Anything I can fix?

**Review by flouthoc — APPROVED** [review]
@Jackbaude thanks LGTM

**Pull request merged**